crates/apollo-compiler/src/database/hir_db.rs

```diff
@@ -253,201 +201,80 @@ fn schema(db: &dyn HirDatabase) -> Arc<SchemaDefinition> {
     Arc::new(schema_def)
 }
 
-fn schema_extensions(db: &dyn HirDatabase) -> Arc<Vec<SchemaExtension>> {
-    let schema_ext = all_type_definitions(db)
-        .filter_map(|(id, definition)| match definition {
-            ast::Definition::SchemaExtension(def) => Some(SchemaExtension {
-                directives: directives(def.directives(), id),
-                root_operation_type_definition: root_operation_type_definition(
-                    def.root_operation_type_definitions(),
-                    id,
-                ),
-                loc: location(id, def.syntax()),
-            }),
-            _ => None,
-        })
-        .collect();
-    Arc::new(schema_ext)
-}
-
-fn object_types(db: &dyn HirDatabase) -> Arc<Vec<ObjectTypeDefinition>> {
-    let objects = all_type_definitions(db)
-        .filter_map(|(id, definition)| match definition {
-            ast::Definition::ObjectTypeDefinition(obj_def) => object_type_definition(obj_def, id),
-            _ => None,
-        })
-        .collect();
-    Arc::new(objects)
-}
-
-fn object_type_extensions(db: &dyn HirDatabase) -> Arc<Vec<ObjectTypeExtension>> {
-    let objects = all_type_definitions(db)
-        .filter_map(|(id, definition)| match definition {
-            ast::Definition::ObjectTypeExtension(def) => Some(ObjectTypeExtension {
-                directives: directives(def.directives(), id),
-                name: name(def.name(), id)?,
-                implements_interfaces: implements_interfaces(def.implements_interfaces(), id),
-                fields_definition: fields_definition(def.fields_definition(), id),
-                loc: location(id, def.syntax()),
-            }),
-            _ => None,
-        })
-        .collect();
-    Arc::new(objects)
-}
-
-fn scalars(db: &dyn HirDatabase) -> Arc<Vec<ScalarTypeDefinition>> {
-    let scalars = all_type_definitions(db)
-        .filter_map(|(id, definition)| match definition {
-            ast::Definition::ScalarTypeDefinition(scalar_def) => scalar_definition(scalar_def, id),
-            _ => None,
-        })
-        .collect();
-    let scalars = built_in_scalars(scalars);
-
-    Arc::new(scalars)
+macro_rules! by_name {
+    ($db: ident, $convert: expr) => {{
+        let mut map = IndexMap::new();
+        for def in type_definitions($db, $convert) {
+            let name = def.name().to_owned();
+            map.entry(name).or_insert_with(|| Arc::new(def));
```

Comment:
we should also record duplicates here for validation. I can work on that since validation isn't the most pressing thing for the router, but then a todo comment would be nice 😇 . probably we could have a struct that contains both the IndexMap and a Vec of errored definitions / extensions.

Reply:
There was already validation code looking for duplicate names. In this PR I’ve changed it to be based on AST definitions where duplicates are still present despite this HIR change.

(BTW it looks like we should check for name collisions of type definitions regardless of their kind, not just collisions among object types, and separately collisions among enums, etc. I can make another PR for that.)

Reply:
Oh I actually read that AST-based validation and then forgot about it when I read this. that's OK, I filed https://github.com/apollographql/apollo-rs/issues/386 for different-kinded collisions, I'd be happy to do it too.